persistence/rocksdb/src/main/java/org/infinispan/persistence/rocksdb/RocksDBStore.java

```diff
@@ -0,0 +1,532 @@
+import org.infinispan.persistence.rocksdb.logging.Log;
+import org.infinispan.persistence.spi.AdvancedLoadWriteStore;
+import org.infinispan.persistence.spi.InitializationContext;
+import org.infinispan.persistence.spi.PersistenceException;
+import org.infinispan.util.logging.LogFactory;
+import org.rocksdb.CompressionType;
+import org.rocksdb.Options;
+import org.rocksdb.ReadOptions;
+import org.rocksdb.RocksDB;
+import org.rocksdb.RocksDBException;
+import org.rocksdb.RocksIterator;
+
+@ConfiguredBy(RocksDBStoreConfiguration.class)
+public class RocksDBStore implements AdvancedLoadWriteStore {
+    private static final Log log = LogFactory.getLog(RocksDBStore.class, Log.class);
+    private RocksDBStoreConfiguration configuration;
+    private BlockingQueue<ExpiryEntry> expiryEntryQueue;
+    private RocksDB db;
+    private RocksDB expiredDb;
+    private InitializationContext ctx;
+    private Semaphore semaphore;
+    private volatile boolean stopped = true;
+
+    @Override
+    public void init(InitializationContext ctx) {
+        this.configuration = ctx.getConfiguration();
+        this.ctx = ctx;
+        this.semaphore = new Semaphore(Integer.MAX_VALUE, true);
+    }
+
+    @Override
+    public void start() {
+        expiryEntryQueue = new LinkedBlockingQueue<>(configuration.expiryQueueSize());
+
+        try {
+            db = openDatabase(getQualifiedLocation(), dataDbOptions());
+            expiredDb = openDatabase(getQualifiedExpiredLocation(), expiredDbOptions());
+            stopped = false;
+        } catch (Exception e) {
+            throw new CacheConfigurationException("Unable to open database", e);
+        }
+    }
+
+    private String sanitizedCacheName() {
+        return ctx.getCache().getName().replaceAll("[^a-zA-Z0-9-_\\.]", "_");
+    }
+
+    private String getQualifiedLocation() {
+        return configuration.location() + sanitizedCacheName();
+    }
+
+    private String getQualifiedExpiredLocation() {
+        return configuration.expiredLocation() + sanitizedCacheName();
+    }
+
+    private Options dataDbOptions() {
+        Options options = new Options().setCreateIfMissing(true);
+        options.setCompressionType(CompressionType.getCompressionType(configuration.compressionType().toString()));
+
+        return options;
+    }
+
+    private Options expiredDbOptions() {
+        return new Options().setCreateIfMissing(true);
+    }
+
+    /**
+     * Creates database if it doesn't exist.
+     */
+    protected RocksDB openDatabase(String location, Options options) throws IOException, RocksDBException {
+        File dir = new File(location);
+        dir.mkdirs();
+        return RocksDB.open(options, location);
+    }
+
+    protected void destroyDatabase(String location) throws IOException {
+        File dir = new File(location);
+        // Rocks DB JNI does not remove files correctly on Windows - so let's do a clean up manually
```

Comment:
Since RocksDB doesn't even have an API for removing a database, we have to do the manual removal dance anyway. I've updated the comment.